labs/SAW/SAW.md

```diff
@@ -41,30 +41,106 @@ tools. As well, there are some examples
 and a full-featured tutorial on SAW
 [here](https://saw.galois.com/intro/).
 
+# Why Use SAW?
+
+Previous labs demonstrated how to look at algorithm specifications
+from math papers and formally represent them in Cryptol.  A
+software/hardware developer now can create a performant implementation
+with confidence since they are armed with a bit-precise, unambiguous,
+machine-readable, formally verified Cryptol specification.
+
+Of course, any developer desires that the code they write completely
+adheres to all of the cryptographic properties pertaining to the
+target algorithm. They also want to uphold best coding practices for
+their implementation language and avoid vulnerabilities such as buffer
+or integer overflows. After all, what good is cryptography if an
+attacker can take advantage of an implementation flaw?
+
+This is where the Software Analysis Workbench (SAW) comes in. SAW is a
+tool that formally verifies properties of compiled code using
+[automated reasoning
+tools](https://www.amazon.science/blog/a-gentle-introduction-to-automated-reasoning). Alright,
+that's a lot of fancy words, but essentially SAW can, in a fairly
+push-button way, verify that, for all inputs, a function does what
+it's supposed to do and nothing more. This makes SAW a much more
+powerful tool for assuring code compared to unit tests, which are
+limited in their testing scope and cannot test for all possible,
+potentially dangerous edge cases.
+
+In order to get SAW to work its magic, source code needs to be
+compiled - SAW works on low-level implementation languages such as
+[LLVM bitcode](https://en.wikipedia.org/wiki/LLVM), [Java byte
+code](https://en.wikipedia.org/wiki/Java_bytecode), low-level Verilog,
+and even some assembly languages. The examples in this lab demonstrate
+verification of cryptographic implementations written in the C
+language, so below you'll see C source code being compiled to LLVM
+bitcode.
+
+Next, we need to tell SAW exactly what to verify. For every

```

Comment:
We don't think about it much when endeavoring for functions to be side-effect free (you know, _functions_), but in other cases a function needs to read/write globals -- so state is essentially an input and an output -- as are pointer arguments.  So a contract could be viewed as a property stating preconditions (for state and arguments) and postconditions (for state, arguments, and the return value).

Reply:
Good point. State has been added to the list of things SAW needs to know.